### PR TITLE
[CIS-693] Channel watcher events

### PR DESCRIPTION
**Pull request by b-onc**
WARNING: it's not tested in a real environment (with 2 devices and such)

### CODE REVIEW

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
+//
+//  UserWatchingEventMiddleware.swift
+//  StreamChat
+//
+//  Created by Bahadir Oncel on 05/03/2021.
+//  Copyright © 2021 Stream.io Inc. All rights reserved.
+//
```
Comment (VojtaStavik) [inline Sources/StreamChat/WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware.swift:7]: Seems like @DominikBucher12's fix broke the formatter 🤷‍♂️

Comment (b-onc) [reply Sources/StreamChat/WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware.swift:7]: o.m.g...

File: Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+/// `ChatChannelWatcherListController` uses this protocol to communicate changes to its delegate.
+///
+/// This protocol can be used only when no custom extra data are specified. If you're using custom extra data types,
+/// please use `_ChatChannelWatcherListControllerDelegate` instead.
+///
+public protocol ChatChannelWatcherListControllerDelegate: DataControllerStateDelegate {
+    /// The controller observed a change in the channel watcher list.
+    func channelWatcherListController(
+        _ controller: ChatChannelWatcherListController,
+        didChangeWatchers watchers: Set<ChatUser>
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift:173]: -> `didChangeWatchers changes: [ListChange<ChatUser>]`

**Review by VojtaStavik — COMMENTED** [review]
I left quite a few comments but the overall direction looks good 👍

File: Sources/StreamChat/WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware_Tests.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
+//
+//  UserWatchingEventMiddleware_Tests.swift
+//  StreamChatTests
+//
+//  Created by Bahadir Oncel on 05/03/2021.
+//  Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/StreamChat/WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware_Tests.swift:8]: Probably this was left from when the formatter didn't work. What's the way to quickly re-format commits?

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 4]
```diff
@@ -258,7 +272,7 @@ extension _ChatChannel {
             isFrozen: dto.isFrozen,
             members: Set(members),
             currentlyTypingMembers: Set(typingMembers),
-            watchers: [],
+            watchers: Set(dto.watchers.map { $0.asModel() }),
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:275]: this is the same issue with `members`, I believe we should add this to that ticket and resolve it together

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift [hunk 5]
```diff
@@ -158,6 +161,17 @@ extension NSManagedObjectContext {
...
+        if let watchers = payload.watchers {
+            // We don't call `removeAll` on watchers since user could've requested
+            // a different page
+            try watchers.forEach {
+                let user = try saveUser(payload: $0)
+                dto.watchers.insert(user)
+            }
+        }
+        // We don't reset `watchers` array if it's missing
+        // since that can mean that user didn't request watchers
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:173]: OK, empty values are actually sent. I've introduced resetting watchers in `ChannelUpdater` where I have the query info so I can know if empty watchers mean empty page or no watchers

File: Sources/StreamChat/Workers/ChannelUpdater.swift [hunk 6]
```diff
@@ -220,4 +220,34 @@ class ChannelUpdater<ExtraData: ExtraDataTypes>: Worker {
...
+    ///
+    /// - Parameters:
+    ///   - query: Query object for watchers. See `ChannelWatcherListQuery`
+    ///   - completion: Called when the API call is finished. Called with `Error` if the remote update fails.
+    func channelWatchers(query: ChannelWatcherListQuery, completion: ((Error?) -> Void)? = nil) {
+        apiClient.request(endpoint: .channelWatchers(query: query)) { (result: Result<ChannelPayload<ExtraData>, Error>) in
+            do {
+                let payload = try result.get()
+                self.database.write { (session) in
+                    if query.pagination.offset == nil || query.pagination.offset == 0 {
```
Comment (VojtaStavik) [inline Sources/StreamChat/Workers/ChannelUpdater.swift:236]: What's the semantic difference between `nil` and `0`? Maybe the offset shouldn't be optional?

File: Sources/StreamChat/Workers/ChannelUpdater.swift [hunk 7]
```diff
@@ -220,4 +220,34 @@ class ChannelUpdater<ExtraData: ExtraDataTypes>: Worker {
...
+                        // This is the first page of the watchers.
+                        // We can clear the existing watchers safely
+                        let channel = session.channel(cid: query.cid)
+                        channel?.watchers.removeAll()
```
Comment (VojtaStavik) [inline Sources/StreamChat/Workers/ChannelUpdater.swift:240]: Let's do this only if the watchers from the payload are empty, too

File: Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    /// - Parameter completion: Called when the API call is finished. Called with `Error` if the remote update fails.
+    func stopWatching(completion: ((Error?) -> Void)? = nil) {
+        // We have no way to check if channel exists on backend
+        // so we'll let backend return an error if it's not
+        updater.stopWatching(cid: query.cid) { error in
+            self.callback {
+                completion?(error)
+            }
+        }
+    }
```
Comment (VojtaStavik) [inline Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift:225]: So we want to expose those both on `ChannelController` and here?? What's the reasoning behind this?

**Review by VojtaStavik — APPROVED** [review]
OK, looks great 👍  I left some comments and questions though.

File: Sources/StreamChat/Database/DTOs/ChannelDTO.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/StreamChat/Database/DTOs/ChannelDTO.swift:275]: CIS-727 created

File: Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift (hunk 8 above)
Comment (b-onc) [reply Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift:225]: I guess we can remove these from here... no reason after we introduced the watcherlistcontroller

Comment (VojtaStavik) [reply Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift:225]: Yep, I think these should live just on `ChannelController`

File: Sources/StreamChat/Workers/ChannelUpdater.swift (hunk 6 above)
Comment (b-onc) [reply Sources/StreamChat/Workers/ChannelUpdater.swift:236]: We don't encode `nil` and backend assumes 0 in that case

File: Package.swift [hunk 9]
```diff
@@ -64,7 +64,6 @@ var streamChatSourcesExcluded: [String] { [
     "WebSocketClient/EventMiddlewares/ChannelTruncatedEventMiddleware_Tests.swift",
     "WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware_Tests.swift",
     "WebSocketClient/EventMiddlewares/MessageReactionsMiddleware_Tests.swift",
-    "WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware_Tests.swift",
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Package.swift]: resolved

File: Package.swift [hunk 10]
```diff
@@ -64,6 +64,7 @@ var streamChatSourcesExcluded: [String] { [
     "WebSocketClient/EventMiddlewares/ChannelTruncatedEventMiddleware_Tests.swift",
     "WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware_Tests.swift",
     "WebSocketClient/EventMiddlewares/MessageReactionsMiddleware_Tests.swift",
+    "WebSocketClient/EventMiddlewares/UserWatchingEventMiddleware_Tests.swift",
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Package.swift:68]: resolved

File: Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift [hunk 11]
```diff
@@ -2142,6 +2142,166 @@ class ChannelController_Tests: StressTestCase {
...
+            completionCalledError = $0
+        }
+        
+        // Simulate failed update
+        let testError = TestError()
+        env.channelUpdater!.stopWatching_completion?(testError)
+        
+        // Completion should be called with the error
+        AssertAsync.willBeEqual(completionCalledError as? TestError, testError)
+    }
```
[the remark this answers is not in the document]
Comment (b-onc) [reply Sources/StreamChat/Controllers/ChannelController/ChannelController_Tests.swift:2304]: ¯\_(ツ)_/¯

File: Sources/StreamChat/Workers/ChannelUpdater.swift (hunk 7 above)
Comment (b-onc) [reply Sources/StreamChat/Workers/ChannelUpdater.swift]: done

File: Sources/StreamChat/Workers/ChannelUpdater_Tests.swift [hunk 12]
```diff
@@ -728,4 +728,77 @@ class ChannelUpdater_Tests: StressTestCase {
...
+        // Create a dummy channel
+        let cid = ChannelId.unique
+        try database.createChannel(cid: cid, withMessages: false)
+        
+        var channel: ChatChannel? {
+            database.viewContext.channel(cid: cid)?.asModel()
+        }
+        
+        // Assert that the dummy channel has a watcher
+        assert(!(channel?.watchers.isEmpty ?? true))
```
Comment (VojtaStavik) [inline Sources/StreamChat/Workers/ChannelUpdater_Tests.swift:784]: nit: You can do just `assert(channel?.watchers.isEmpty != true)` which is maybe a bit more readable :)

**Review by VojtaStavik — APPROVED** [review]
🎉

**Pull request merged**